src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        if (typeof action.extraData === 'object' && action.extraData !== null) {
+          Object.assign(currentState, action.extraData);
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+      default:
```

Comment:
The switch statement is well implemented with a default case for error handling. However, you should ensure that the default case provides meaningful feedback or handles unexpected action types appropriately.